File: infrastructures/ddd/typescript/src/api/middleware/OwnershipMiddleware.ts
import { Request, Response, NextFunction } from 'express';

/**
 * Middleware for verifying resource ownership.
 * MUST be applied AFTER authentication but BEFORE UnitOfWork.
 * 
 * This middleware:
 * - Extracts user ID from authenticated request
 * - Extracts resource ID from request params
 * - Verifies user owns the resource
 * - Returns 403 Forbidden if ownership check fails
 * 
 * For public resources, skip this middleware on those routes.
 * 
 * Middleware order:
 * 1. Correlation ID → 2. Logging → 3. Authentication → 
 * 4. Authorization/Ownership (THIS) → 5. UnitOfWork → 6. Controller →
 * 7. UnitOfWork commit/rollback → 8. Error Handling
 */
export class OwnershipMiddleware {
  /**
   * Service for checking resource ownership.
   * In a real implementation, inject a service that queries database.
   */
  private readonly _ownershipChecker: (userId: string, resourceId: string) => Promise<boolean>;

  constructor(p_ownershipChecker: (userId: string, resourceId: string) => Promise<boolean>) {
    this._ownershipChecker = p_ownershipChecker;
  }

  /**
   * Middleware handler function.
   * Use with Express: app.use((req, res, next) => middleware.Handle(req, res, next))
   */
  public async Handle(
    p_request: Request,
    p_response: Response,
    p_next: NextFunction
  ): Promise<void> {
    try {
      // Extract user ID from authenticated request
      const userId = (p_request as any).user?.id;
      if (!userId) {
        p_response.status(401).json({ error: 'Unauthorized' });
        return;
      }

      // Extract resource ID from params (e.g., /api/v1/example/:id)
      const resourceId = p_request.params.id;
      if (!resourceId) {
        // No resource ID in params, skip ownership check
        p_next();
        return;
      }

      // Check if resource is public (skip ownership check)
      // In a real implementation, check against a whitelist or resource metadata
      const isPublicResource = false; // Example: await this._isPublicResource(resourceId)
      if (isPublicResource) {
        p_next();
        return;
      }

      // Verify ownership
      const ownsResource = await this._ownershipChecker(userId, resourceId);
      if (!ownsResource) {
        p_response.status(403).json({ error: 'Access denied - you do not own this resource' });
        return;
      }

      p_next();
    } catch (error) {
      p_next(error);
    }
  }
}

File: infrastructures/ddd/python/src/api/middleware/ownership_middleware.py
"""
Middleware for validating resource ownership.
MUST be applied AFTER authentication but BEFORE controller execution.
"""

from typing import Awaitable, Callable


class OwnershipMiddleware:
    """
    Middleware for ownership validation.
    
    This middleware:
    - Extracts user ID from auth context
    - Extracts resource ID from request
    - Verifies user owns the resource
    - Returns 403 Forbidden if not owned
    
    FastAPI implementation:
    ```python
    from fastapi import Request, HTTPException, status
    from starlette.middleware.base import BaseHTTPMiddleware
    
    class OwnershipMiddleware(BaseHTTPMiddleware):
        def __init__(self, app, ownership_service):
            super().__init__(app)
            self._ownership_service = ownership_service
        
        async def dispatch(self, request: Request, call_next):
            # Skip for public resources or non-mutating operations
            if request.method == "GET" or request.url.path.endswith("/public"):
                return await call_next(request)
            
            # Extract user ID from auth context
            user_id = request.state.user_id
            
            # Extract resource ID from path
            resource_id = request.path_params.get("id")
            
            if resource_id:
                # Verify ownership
                if not await self._ownership_service.verify_async(user_id, resource_id):
                    raise HTTPException(
                        status_code=status.HTTP_403_FORBIDDEN,
                        detail="Access denied"
                    )
            
            return await call_next(request)
    ```
    
    Flask implementation:
    ```python
    from flask import request, g, abort
    
    def ownership_middleware():
        # Skip for GET requests
        if request.method == "GET":
            return
        
        # Get user ID from g
        user_id = g.user_id
        
        # Get resource ID from URL
        resource_id = request.view_args.get("id")
        
        if resource_id:
            # Verify ownership
            if not ownership_service.verify(user_id, resource_id):
                abort(403, "Access denied")
    ```
    """
    
    async def __call__(
        self,
        p_request: any,
        p_call_next: Callable[[any], Awaitable[any]]
    ) -> any:
        """
        Process the request and validate ownership.
        
        Args:
            p_request: The HTTP request
            p_call_next: The next middleware/handler
            
        Returns:
            The HTTP response
            
        Raises:
            HTTPException: If ownership validation fails
        """
        # Skip for GET requests (read-only)
        if p_request.method == "GET":
            return await p_call_next(p_request)
        
        # Extract user ID (from auth context)
        # user_id = p_request.state.user_id
        
        # Extract resource ID (from path params)
        # resource_id = p_request.path_params.get("id")
        
        # Verify ownership if resource ID present
        # if resource_id:
        #     if not await self._verify_ownership(user_id, resource_id):
        #         raise HTTPException(403, "Access denied")
        
        return await p_call_next(p_request)
